### PR TITLE
Adding permissions and more routes

### PR DESCRIPTION
**Pull request by mikefranze**
Adding permissions based on owner, admin, and auditor status
Adding a few more routes to backend.

### CODE REVIEW

**Review by ArendPeter — APPROVED** [review]

File: backend/src/Controllers/auth.controllers.ts [hunk 1]
```diff
@@ -1,17 +1,29 @@
...
 
 var jwt = require('jsonwebtoken')
 
 const getUser = (req: any, res: any, next: any) => {
   console.log('-> auth.getUser')
   //This will not verify whether the signature is valid, need to add verification
   req.user = jwt.decode(req.cookies.id_token)
+  req.user_auth = {}
+  if (req.user && req.election){
+    req.user_auth.isOwner = req.user.sub === req.election.owner_id
```
[the remark this answers is not in the document]
Comment (mikefranze) [reply backend/src/Controllers/auth.controllers.ts:12]: I went back and forth on this not sure the best way to do it. I might revisit it later with your suggestions. I might also merge our existing field voter_auth into user_auth which has info on if they're authorized and if they've already voted. Then we could add registration status to that too.

**Comment by scottPlusPlus** [comment]
Is there a reason to add all of these as express middleware, rather than just functions?

File: backend/src/Controllers/auth.controllers.ts [hunk 2]
```diff
@@ -32,8 +44,222 @@ const assertOwnership = (req: any, res: any, next: any) => {
...
+    return res.status('400').json({
+      error: "User is not authorized to edit election auditors"
+    })
+  }
+  next()
+}
+
+const canViewElection = (req: any, res: any, next: any) => {
+  console.log(`-> auth.canEditElection`)
+  if (!(req.user_auth.isOwner || req.user_auth.isAdmin || req.user_auth.isSystemAdmin)) {
```
[the remark this answers is not in the document]
Comment (mikefranze) [reply backend/src/Controllers/auth.controllers.ts:73]: Auditor was left out of this permission. They're main responsibility is review.

File: backend/src/Controllers/auth.controllers.ts [hunk 3]
```diff
@@ -32,8 +44,222 @@ const assertOwnership = (req: any, res: any, next: any) => {
...
+      error: "User is not authorized to view ballots"
+    })
+  }
+  next()
+}
+
+const canViewBallot = (req: any, res: any, next: any) => {
+  console.log(`-> auth.canViewBallot`)
+  //TODO: add permission for voter to view their own ballot
+  if (!(req.user_auth.isSystemAdmin)) {
```
[the remark this answers is not in the document]
Comment (mikefranze) [reply backend/src/Controllers/auth.controllers.ts:174]: I think admins and auditors can view all ballots, but this is for a user looking up their ballot based on their res.user info.

File: backend/src/Controllers/auth.controllers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArendPeter) [reply backend/src/Controllers/auth.controllers.ts:73]: image

Ah ok, does it need to be updated in the spreadsheet then?

**Comment by mikefranze** [comment]
> Is there a reason to add all of these as express middleware, rather than just functions?

I went back and forth on this too, I thought it might be better to have the permission check be separated from logic following the check. That being said this could probably be condensed down to just a single middleware.

File: backend/src/Controllers/auth.controllers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikefranze) [reply backend/src/Controllers/auth.controllers.ts:73]: Ah, eyes skipped over the actual permission and just looked at the console log which was for editing the election. Both should be fixed.

**Comment by mikefranze** [comment]
New commit simplifies this a lot.

**Pull request merged**